### DISCUSSION TITLE
"asSeenFrom" in Scala 3

### DISCUSSION THREAD

**Discussion by xuwei-k** [scala/scala3#14957] [Metaprogramming]
I need `asSeenFrom` method similar to Scala 2 reflection API. but I could not found in Scala 3 API.

https://github.com/scala/scala/blob/18ce8d410b57f30f4161896b49b03d2d3e7b0401/src/reflect/scala/reflect/api/Types.scala#L246

`dotty.tools.dotc.core.TypeOps` has `asSeenFrom` but I think we can not use that from quotes API.

https://github.com/lampepfl/dotty/blob/53f5531e7e31eca02b3db158c8c384e713b4542a/compiler/src/dotty/tools/dotc/core/TypeOps.scala#L30

Is there more better solution to avoid compiler internal api.

please see following example.

https://github.com/xuwei-k/Scala-asSeenFrom/commit/830080616242047e3af8115ff97a82f5e4cb65d6

Thanks.

**Comment by smarter** [comment]
Does this work for your usecase?
https://github.com/lampepfl/dotty/blob/53f5531e7e31eca02b3db158c8c384e713b4542a/library/src/scala/quoted/Quotes.scala#L2590-L2594
(It's implemented using asSeenFrom: https://github.com/lampepfl/dotty/blob/53f5531e7e31eca02b3db158c8c384e713b4542a/compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala#L1740-L1741)

**Comment by xuwei-k** [comment]
`memberType` does not work in my usecase. because it fixed `cls = member.owner` param.

**Reply by smarter** [reply]
Can you give a concrete example of what output you expect and what output you get with memberType?

**Reply by xuwei-k** [reply]
> please see following example.
>
> https://github.com/xuwei-k/Scala-asSeenFrom/commit/830080616242047e3af8115ff97a82f5e4cb65d6


🙏 

```scala
class MyAnnotation(clazz: Class[?]) extends StaticAnnotation

trait D[E]

class A[B] {
  @MyAnnotation(clazz = classOf[D[B]])
  def x: B = ???
}

class C extends A[Int]
```


I expect `D[Int]`.

https://github.com/xuwei-k/Scala-asSeenFrom/commit/64784c7b31fb1b36db2e8231101573b00a990cf8

memberType result

```scala
ClassInfo(ThisType(TypeRef(NoPrefix,module class example)), class C, List(AppliedType(TypeRef(ThisType(TypeRef(NoPrefix,module class example)),class A),List(TypeRef(TermRef(ThisType(TypeRef(NoPrefix,module class <root>)),object scala),Int)))))
```

```scala
ClassInfo(ThisType(TypeRef(NoPrefix,module class example)), class A, List(TypeRef(ThisType(TypeRef(NoPrefix,module class lang)),class Object)))
```

**Reply by smarter** [reply]
Ah I see, you're interested in the type of' `D[B]` in the scope of `C`, indeed it seems like we're missing an API for doing that /cc @nicolasstucki, you could open an issue for this (and maybe even a PR with a proposed API addition).
